ethereum/contracts/relayer/coreRelayer/CoreRelayerDelivery.sol

```diff
@@ -1,623 +1,560 @@
+    //same chain refund
+    if (refundChainId == getChainId())
+      return pay(payable(fromWormholeFormat(refundAddress)), refundAmount)
+        ? RefundStatus.REFUND_SENT
+        : RefundStatus.REFUND_FAIL;
+    
+    //cross-chain refund
+    uint256 wormholeMessageFee = getWormhole().messageFee();
+    if (refundAmount <= wormholeMessageFee)
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_NOT_ENOUGH;
+
+    IRelayProvider relayProvider = IRelayProvider(fromWormholeFormat(relayerAddress));
+    if (!relayProvider.isChainSupported(refundChainId))
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_PROVIDER_NOT_SUPPORTED;
+
+    uint256 refundSubMessageFee;
+    unchecked{refundSubMessageFee = refundAmount - wormholeMessageFee;}
+
+    uint256 receiverValueTarget = calculateTargetDeliveryMaximumRefund(
+      refundChainId, refundSubMessageFee, relayProvider
+    );
+    
+    //If refundAmount is not enough to pay for one wei of receiver value, then do not perform the
+    //  cross-chain refund (i.e. if (delivery overhead) + (wormhole message fee) + (cost of one wei
+    //  of receiver value) is larger than the remaining refund)
+    //TODO AMO: but what happens to the value then?
+    if (receiverValueTarget == 0)
+      return RefundStatus.CROSS_CHAIN_REFUND_FAIL_NOT_ENOUGH;
+
+    uint256 maxBudget = relayProvider.quoteMaximumBudget(refundChainId);
+    bool exceedsMaxBudget = false;
+    if (receiverValueTarget > maxBudget) {
+      //TODO AMO: What happens with the difference if the maximum budget is exceeded?
+      receiverValueTarget = maxBudget;
+      exceedsMaxBudget = true;
     }
 
-    /**
-     * @notice vaaKeysWithVAAs checks that signedVaa matches the description given by 'vaaKey'
-     * Specifically, if 'vaaKey.infoType' is VaaKeyType.EMITTER_SEQUENCE, then we check
-     * if the emitterAddress and sequence match
-     * else if 'vaaKey.infoType' is VaaKeyType.EMITTER_SEQUENCE, then we check if the VAA hash matches
-     *
-     * @param vaaKey VaaKey struct describing a wormhole message (VAA)
-     * @param signedVaa signed wormhole message
-     */
-    function vaaKeyMatchesVAA(
-        IWormholeRelayer.VaaKey memory vaaKey,
-        bytes memory signedVaa
-    ) internal view returns (bool) {
-        IWormhole.VM memory parsedVaa = wormhole().parseVM(signedVaa);
-        if (vaaKey.infoType == IWormholeRelayer.VaaKeyType.EMITTER_SEQUENCE) {
-            return (vaaKey.chainId == parsedVaa.emitterChainId)
-                && (vaaKey.emitterAddress == parsedVaa.emitterAddress)
-                && (vaaKey.sequence == parsedVaa.sequence);
-        } else if (vaaKey.infoType == IWormholeRelayer.VaaKeyType.VAAHASH) {
-            return (vaaKey.vaaHash == parsedVaa.hash);
-        } else {
-            return false;
-        }
-    }
+    publishAndPay(
+      wormholeMessageFee,
+      0,
+      refundSubMessageFee,
+      //TODO AMO: Why does this not use convertSendToDeliveryInstruction()?
```

Comment:
Now it does with https://github.com/wormhole-foundation/wormhole/pull/2898 ! good callout :)